From 4da66d3e5fdf15ad9abdebbba4a86b97b7fd1aa2 Mon Sep 17 00:00:00 2001
From: Karthikeyan K <karthikk@habana.ai>
Date: Sat, 03 Sep 2022 12:00:52 +0300
Subject: [PATCH] [SW-99831] gaudi3: d2d: implement dphy hw errata fix

This patch implements the sw workaround for GAUDI3_0314.

Problem:
Hardware has a dedicated FSM which is responsible for initializing
the D2D. This initialization FSM has several steps, each step has
a watchdog timer (WD). The default values of the WD timer assumes
200MHz clock while the actual clock in recovery is 50MHz.

Implication:
The WD timer will expire later which will make the initialization
process longer.

Change-Id: I3844f8f3847ebae01baf00df8e6dbc036240df23
---

diff --git a/arch/gaudi3/arc/recovery.c b/arch/gaudi3/arc/recovery.c
index 1ed92fc..235d6d3 100644
--- a/arch/gaudi3/arc/recovery.c
+++ b/arch/gaudi3/arc/recovery.c
@@ -145,6 +145,22 @@
 	}
 }
 
+#define PLL_LOCK_TIMER_VAL		0xC350
+#define OFFSET_CALLOCK_TIMER		0xC350
+#define TIMER_1US			0x32
+#define NUM_OF_PHY			6
+#define INTER_PHY_DIST			(mmD0_DPHY1_CTRL_BASE - mmD0_DPHY0_CTRL_BASE)
+static void arcboot_rom_d2d_phy_wd_hw_errata_fix(void)
+{
+	uint32_t i;
+
+	for (i = 0; i < NUM_OF_PHY; i++) {
+		regmap_writel(PLL_LOCK_TIMER_VAL, mmD0_DPHY0_CTRL_INIT_PLL_CTRL0 + i * INTER_PHY_DIST);
+		regmap_writel(OFFSET_CALLOCK_TIMER, mmD0_DPHY0_CTRL_INIT_PLL_CTRL1 + i * INTER_PHY_DIST);
+		regmap_writel(TIMER_1US, mmD0_DPHY0_CTRL_INIT_PLL_CTRL3 + i * INTER_PHY_DIST);
+	}
+}
+
 #define D2D_LINK_STATUS_UDELAY		10000 /* 10 ms */
 static void rom_d2d_link_init(void)
 {
@@ -167,6 +183,11 @@
 		return;
 	}
 
+	/*
+	 * SW WA for HW Errata GAUDI3_0314
+	 */
+	arcboot_rom_d2d_phy_wd_hw_errata_fix();
+
 	/* Set START bit for D2D link initialization */
 	regmap_writel(D0_PSOC_GLOBAL_CONF_DPHY_FSM_CTRL_START_M, mmD0_PSOC_GLOBAL_CONF_DPHY_FSM_CTRL);
 
